File: 2023/python/day01.py
#!/usr/bin/env python3

import sys
import re

p1 = p2 = 0
names = ["zero", "one", "two", "three", "four", "five", "six", "seven",
         "eight", "nine"]
re2 = "(?=(" + "|".join(names) + "|[0-9]))"
map = {}

for i, name in enumerate(names):
  map[name] = str(i)
for i in range(10):
  map[str(i)] = str(i)


for line in sys.stdin.readlines():
  digits = re.findall("[0-9]", line)
  p1 += int(digits[0] + digits[-1])
  digits = re.findall(re2, line)
  p2 += int(map[digits[0]] + map[digits[-1]])

print("01:", p1, p2)

File: 2023/kotlin/src/Day01.kt
import java.io.File

fun toValue(l: String, p2: Boolean): Int {
    val dig = l.indices
        .mapNotNull {
            when {
                l[it].isDigit() -> l[it]
                p2 && l.startsWith("zero", it) -> '0'
                p2 && l.startsWith("one", it) -> '1'
                p2 && l.startsWith("two", it) -> '2'
                p2 && l.startsWith("three", it) -> '3'
                p2 && l.startsWith("four", it) -> '4'
                p2 && l.startsWith("five", it) -> '5'
                p2 && l.startsWith("six", it) -> '6'
                p2 && l.startsWith("seven", it) -> '7'
                p2 && l.startsWith("eight", it) -> '8'
                p2 && l.startsWith("nine", it) -> '9'
                else -> null
            }
        }
        .toList()

    return ("" + dig.first() + dig.last()).toInt()
}

fun main(args: Array<String>) {
    val lines = File(args.getOrNull(0) ?: "../../inputs/2023/01-input.txt").readLines()
    val p1 = lines.sumOf { toValue(it, false) }
    val p2 = lines.sumOf { toValue(it, true) }
    println("01: $p1 $p2")
}
